substrate/frame/scheduler/src/lib.rs

```diff
@@ -1157,24 +1190,51 @@ impl<T: Config> Pallet<T> {
 			return
 		}
 
-		let mut incomplete_since = now + One::one();
-		let mut when = IncompleteSince::<T>::take().unwrap_or(now);
-		let mut executed = 0;
+		let queue = Queue::<T>::get();
+		let end_index = match queue.iter().position(|&x| x >= now) {
+			Some(end_index) => end_index.saturating_add(1),
+			None => {
+				return;
+			},
+		};
+		if end_index == 0 {
+			return;
+		}
 
-		let max_items = T::MaxScheduledPerBlock::get();
-		let mut count_down = max;
-		let service_agenda_base_weight = T::WeightInfo::service_agenda_base(max_items);
-		while count_down > 0 && when <= now && weight.can_consume(service_agenda_base_weight) {
-			if !Self::service_agenda(weight, &mut executed, now, when, u32::max_value()) {
-				incomplete_since = incomplete_since.min(when);
+		let mut index = 0;
+		let queue_len = queue.len();
+		for when in queue.iter().skip(index) {
+			if *when < now.saturating_sub(T::MaxStaleTaskAge::get()) {
+				Agenda::<T>::remove(*when);
+			} else {
+				let mut executed = 0;
+				if !Self::service_agenda(weight, &mut executed, now, *when, max) {
+					break;
+				}
+			}
+			index = index.saturating_add(1);
+			if index >= queue_len {
+				break;
 			}
-			when.saturating_inc();
-			count_down.saturating_dec();
-		}
-		incomplete_since = incomplete_since.min(when);
-		if incomplete_since <= now {
-			IncompleteSince::<T>::put(incomplete_since);
 		}
+
+		Queue::<T>::mutate(|queue| {
+			let mut iter = queue.iter();
+			let mut to_remove = Vec::new(); // Collect items to remove
+
+			// Iterate and collect items to remove
+			for _ in 0..index {
+				iter.next();
+			}
+			for item in iter {
+				to_remove.push(*item);
+			}
+
+			// Now remove the collected items
+			for item in to_remove {
+				queue.remove(&item);
+			}
```

Comment:
Not sure what is going on here, but maybe something like `queue.iter().drain().take(index).map(drop);` would work.

Reply:
If the interface for bounded vec is too annoying you can convert to a vec and convert back. Because the code is difficult to read.